### PR TITLE
feat: fox farming provider -> programmatic UniV2 provider

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

tl;dr, decouples ETH/FOX staking from farming, makes ETH/FOX staking UniV2 programmatic without ETH/FOX mentions by generalizing names and fetching whatever we need on-chain, and moves many things to the right place while at it.

This PR:

- Renames DeFiProvider.FoxFarming (Shapeshift Farming) to DeFiProvider.UniV2 ('Uniswap V2')
- Renames the resolver from `foxFarming` to `uniV2`
- Brings the current `foxFarming` as `ethFoxStaking` to be more specific and to decouple our current mental model of `FoxFarming` being both LP and farming
- Moves contracts to `src/contracts` now that we've decouples UniV2 and ETH/FOX staking
- Makes the whole resolver programmatic, e.g no hardcoded tokens 0 and
  1, these are now gotten directly from the LP contract
- Move the farming resolver outside of the `uniV2` resolver, and
  bring back `FoxFarming`. We will now need to differentiate the two, to
  show UniV2 for LP, and Shapeshift Farming for our own farming contract
  leveraging LP tokens
  
- [ ] Typegen UniV2 ABI ~~Too early for this, we do not yet support other pairs, TBD when we bring zerion SDK in~~

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/blob/develop/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

N/A

## Risk

ETH/FOX LP/Staking could be borked, test accordingly

## Testing

- ETH/FOX LP shows no regressions
- ETH/FOX staking shows no regressions
- ETH/FOX LP is displayed as a "Uniswap V2" opportunity
- FOX Page shows no regressions
- Account/asset pages earn opportunities show no regressions
- ETH/FOX LP price/balance show no regressions

### Engineering

- tsc is happy

### Operations

- ☝🏽 

## Screenshots (if applicable)

image
image

### CODE REVIEW

**Comment by gomesalexandre** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #3863** <a href="https://app.graphite.dev/github/pr/shapeshift/web/3863" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/3863?utm_source=stack-comment).

File: src/state/slices/opportunitiesSlice/resolvers/uniV2/constants.ts [hunk 1]
```diff
@@ -0,0 +1 @@
+export const FOX_TOKEN_CONTRACT_ADDRESS = '0xc770EEfAd204B5180dF6a14Ee197D99d808ee52d'
```
Comment (gomesalexandre) [inline src/state/slices/opportunitiesSlice/resolvers/uniV2/constants.ts:1]: Exclusively used in LP domain 
image

File: src/state/slices/opportunitiesSlice/resolvers/uniV2/index.ts [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+  if (bnOrZero(token0MarketData?.price).isZero()) {
+    throw new Error(`Market data not ready for ${underlyingAssetIds[0]}`)
+  }
+
+  const token0Price = token0MarketData.price
+
+  const { assetReference: contractAddress } = fromAssetId(opportunityId)
+  // TODO(gomes): discrimination required because of typechain
+  // Import the standard UniV2 Pool ABI and cast `contractAddress` with it once we bring in Zerion SDK
```
Comment (gomesalexandre) [inline src/state/slices/opportunitiesSlice/resolvers/uniV2/index.ts:82]: @reallybeard FYI

**Comment by gomesalexandre** [comment]
Opening for early review - will use this PR as an opportunity to consolidate contract addresses / contract `AssetId`s in one place but conceptually this is it

File: src/plugins/foxPage/utils.test.ts [hunk 3]
```diff
@@ -62,7 +74,8 @@ describe('foxpage-utils', () => {
   it('should calculate correct APR from given reserves', async () => {
     const input = {
       blockNumber,
-      uniswapLPContract: mockContract,
+      // uniswapLPContract: mockContract,
```
Comment (gomesalexandre) [inline src/plugins/foxPage/utils.test.ts:77]: rm

**Review by 0xApotheosis — APPROVED** [review]
Sturdy as a mountain goat 🐐 

Did not observe any regressions across FOX Opportunities or the DeFi page.

File: src/state/slices/opportunitiesSlice/resolvers/uniV2/utils.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+import { TRADING_FEE_RATE } from './constants'
+
+export const fetchPairData = memoize(
+  async (
+    tokenA: Token,
+    tokenB: Token,
+    fetchPairData: typeof Fetcher['fetchPairData'],
+    provider: providers.Web3Provider,
+  ) => await fetchPairData(tokenA, tokenB, provider),
+)
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply src/state/slices/opportunitiesSlice/resolvers/uniV2/utils.ts]: Nice catch ser! I like seeing dead code go 👋🏽 
5d7a79df812f8f7ec6251fd57babe299d9b29462

**Review by 0xApotheosis — APPROVED** [review]
